modules/core/src/main/java/org/apache/ignite/internal/util/distributed/DistributedProcess.java

```diff
@@ -0,0 +1,388 @@
+     * @param crd Coordinator node to send message.
+     */
+    private void sendSingleMessage(Process p, ClusterNode crd) {
+        assert p.resFut.isDone();
+
+        SingleNodeMessage<R> singleMsg = new SingleNodeMessage<>(p.id, type.ordinal(), p.resFut.result(),
+            (Exception)p.resFut.error());
+
+        if (crd.isLocal())
+            onSingleNodeMessageReceived(singleMsg, crd.id());
+        else {
+            try {
+                ctx.io().sendToGridTopic(crd, GridTopic.TOPIC_DISTRIBUTED_PROCESS, singleMsg, SYSTEM_POOL);
+            }
+            catch (IgniteCheckedException e) {
+                log.error("Unable to send message to coordinator.", e);
+
+                ctx.failure().process(new FailureContext(CRITICAL_ERROR,
+                    new Exception("Unable to send message to coordinator.", e)));
+            }
+        }
+    }
+
+    /**
+     * Processes the received single node message.
+     *
+     * @param msg Message.
+     * @param nodeId Node id.
+     */
+    private void onSingleNodeMessageReceived(SingleNodeMessage<R> msg, UUID nodeId) {
+        Process p = processes.computeIfAbsent(msg.processId(), id -> new Process(msg.processId()));
+
+        p.initCrdFut.listen(f -> {
+            boolean rmvd, isEmpty;
+
+            synchronized (mux) {
+                rmvd = p.remaining.remove(nodeId);
+
+                isEmpty = p.remaining.isEmpty();
+            }
+
+            if (rmvd) {
+                p.singleMsgs.put(nodeId, msg);
+
+                if (isEmpty)
+                    finishProcess(p);
+            }
+        });
+    }
+
+    /**
+     * Creates and sends finish message when all single nodes result received.
+     *
+     * @param p Process.
+     */
+    private void finishProcess(Process p) {
+        HashMap<UUID, R> res = new HashMap<>();
+
+        HashMap<UUID, Exception> err = new HashMap<>();
+
+        p.singleMsgs.forEach((uuid, msg) -> {
+            if (msg.hasError())
+                err.put(uuid, msg.error());
+            else
+                res.put(uuid, msg.response());
+        });
+
+        FullMessage msg = new FullMessage<>(p.id, type.ordinal(), res, err);
+
+        try {
+            ctx.discovery().sendCustomEvent(msg);
+        }
+        catch (IgniteCheckedException e) {
+            log.warning("Unable to send action message.", e);
+        }
+    }
+
+    /**
+     * Handles case when all server nodes have left the grid.
+     */
+    private void onAllServersLeft() {
+        processes.clear();
+    }
+
+    /** @return Cluster coordinator, {@code null} if failed to determine. */
+    private @Nullable ClusterNode coordinator() {
+        return U.oldest(ctx.discovery().aliveServerNodes(), null);
+    }
+
+    /** */
+    private class Process {
+        /** Process id. */
+        private final UUID id;
+
+        /** Init coordinator future. */
+        private final GridFutureAdapter<Void> initCrdFut = new GridFutureAdapter<>();
```

Comment:
Should this future be re-inited in case of the process coordinator changed?

Reply:
No, each node has its own pre-created crd futures. When the coordinator changes, coordinator node will initialize the init crd future locally.